modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/HectorBasedMessageStoreImpl.java

```diff
@@ -470,20 +476,63 @@ public AndesMessageMetadata getMetaData(long messageId) throws AndesException {
                                                                        long firstMsgId, int count)
             throws AndesException {
 
+        Timer timer = new Timer();
+        List<AndesMessageMetadata> messagesFromQueue = new ArrayList<AndesMessageMetadata>();
+        long lastMsgId;
+
         Context context = MetricManager.timer(Level.DEBUG, MetricsConstants.GET_NEXT_MESSAGE_METADATA_FROM_QUEUE).start();
+
+        if (firstMsgId == 0) {
+            firstMsgId = ServerStartupRecoveryUtils.getStartMessageIdForWarmStartup();
+        }
+        long messageIdDifference = ServerStartupRecoveryUtils.getMessageDifferenceForWarmStartup();
+        lastMsgId = firstMsgId + messageIdDifference;
         try {
-            return HectorDataAccessHelper
+            messagesFromQueue = HectorDataAccessHelper
                     .getMessagesFromQueue(queueName,
                             HectorConstants.META_DATA_COLUMN_FAMILY,
-                            keyspace, firstMsgId, Long.MAX_VALUE,
+                            keyspace, firstMsgId, lastMsgId,
                             count, true);
+            if (messagesFromQueue.size() == 0) {
+                readingCassandraInfoLog(timer);
+                while (lastMsgId <= lastRecoveryMessageId) {
+                    long nextMsgId = lastMsgId + 1;
+                    lastMsgId = nextMsgId + messageIdDifference;
+                    messagesFromQueue = HectorDataAccessHelper
+                            .getMessagesFromQueue(queueName,
+                                    HectorConstants.META_DATA_COLUMN_FAMILY,
+                                    keyspace, nextMsgId, lastMsgId,
+                                    count, true);
+                    if (messagesFromQueue.size() > 0) {
+                        break;
```

Comment:
count parameter used in query to limit the result. query may return number of records which may equal or less than to count. While loop going to execute if first query to the Cassandra return 0 metadata which means it's already read tombstone. There after it's continuously running adding up start message id and end message id until it's meet actual records. If query retrieve some actual record, then there should be way to break the while loop and return that result to caller. Query may return number of messages equal or less than to count at this moment. That is what it's checked 'metadataList.size() > 0'. I can change this to 'metadataList.size() <= count' which is give same effect.